### PR TITLE
[SPARK-37627][SQL][FOLLOWUP] Separate SortedBucketTransform from BucketTransform

### PR DESCRIPTION
**Pull request by huaxingao**
### What changes were proposed in this pull request?

1. Currently only a single bucket column is supported in `BucketTransform`, fix the code to make multiple bucket columns work.
2. Separate `SortedBucketTransform` from `BucketTransform`, and make the `arguments` in `SortedBucketTransform` in the format of `columns numBuckets sortedColumns` so we have a way to find out the `columns` and `sortedColumns`.
3. add more test coverage.

### Why are the changes needed?

Fix bugs in `BucketTransform` and `SortedBucketTransform`.

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
New tests

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/connector/expressions/expressions.scala
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/connector/expressions/expressions.scala:115]: does this really work? I don't see a way for people to extract bucket and sort columns from arguments with the `Transform` API.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/connector/expressions/expressions.scala
Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/connector/expressions/expressions.scala:119]: Shall we keep consistent order of `columns` and `numBuckets` for two cases in `arguments`?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/connector/expressions/expressions.scala
Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/connector/expressions/expressions.scala:158]: There is an extra space between `case` and `Lit`.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/connector/expressions/TransformExtractorSuite.scala
Comment (viirya) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/connector/expressions/TransformExtractorSuite.scala:43]: unnecessary change.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/connector/expressions/expressions.scala
Comment (huaxingao) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/connector/expressions/expressions.scala:119]: If there are `sortedColumn`, we need `numBuckets` in between of `columns` and `sortedColumns`, because we need a way to figure out which elements in the array are for columns, and which elements are for sortedColumns.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/connector/expressions/expressions.scala
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/connector/expressions/expressions.scala:128]: is it: `columns = newReferences.take(columns.length), sortedColumns = newReferences.drop(columns.length)`

File: sql/catalyst/src/main/scala/org/apache/spark/sql/connector/expressions/expressions.scala
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/connector/expressions/expressions.scala:107]: Can we create a new class `SortedBucketTransform`?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/connector/expressions/expressions.scala
Comment (huaxingao) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/connector/expressions/expressions.scala]: Changed. Thanks!

File: sql/catalyst/src/main/scala/org/apache/spark/sql/connector/expressions/expressions.scala
Comment (huaxingao) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/connector/expressions/expressions.scala]: Added a new class `SortedBucketTransform`. Thanks!

File: sql/core/src/test/scala/org/apache/spark/sql/connector/DataSourceV2SQLSuite.scala
Comment (huaxingao) [reply sql/core/src/test/scala/org/apache/spark/sql/connector/DataSourceV2SQLSuite.scala:1570]: Just want to make sure multiple columns/sortedColumns work ok.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/connector/expressions/expressions.scala
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/connector/expressions/expressions.scala:48]: It's OK to keep the name `bucket`, to match the name of this SQL feature

File: sql/catalyst/src/main/scala/org/apache/spark/sql/connector/expressions/expressions.scala
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/connector/expressions/expressions.scala:151]: `sorted_bucket` is more SQL-ish.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/connector/catalog/InMemoryTable.scala
Comment (cloud-fan) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/connector/catalog/InMemoryTable.scala:165]: I think we can have a single `BucketTransform.unapply`, to match both `BucketTransform` and `SortedBucketTransform`, so that we can have a single case here and avoid duplicated code.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/connector/expressions/expressions.scala
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/connector/expressions/expressions.scala:124]: where do we use this `unapply`?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/connector/expressions/expressions.scala
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/connector/expressions/expressions.scala:142]: nit: we can do `arguments.zipWithIndex.foreach`, so that it's much easier to get `posOfLit`.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/connector/expressions/expressions.scala
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/connector/expressions/expressions.scala:149]: we know that the arguments of `bucket`/`sorted_bucket`are all `NamedReference`, how about `arguments.take(posOfLit).map(_.asInstanceOf[NamedReference])`?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/connector/expressions/expressions.scala
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/connector/expressions/expressions.scala:151]: this doesn't seem to be right. It only matches `bucket` with a single bucket column.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/connector/expressions/expressions.scala
Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/connector/expressions/expressions.scala:124]: Could you add some comments on `unapply` (if it is really used) about what it returns?

Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/connector/expressions/expressions.scala:124]: BTW, why `def unapply(expr: Expression)` addresses only `BucketTransform` but `def unapply(transform: Transform)` addresses both `sorted_bucket` and `bucket`?

**Review by viirya — COMMENTED** [review]
This change seems to be not only "Add tests ...". It's better to update the title and description accordingly before merging.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/connector/expressions/expressions.scala
Comment (huaxingao) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/connector/expressions/expressions.scala:124]: This was introduced in https://github.com/apache/spark/pull/30706 but doesn't seem to be used. I will remove for now.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/connector/expressions/expressions.scala
Comment (huaxingao) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/connector/expressions/expressions.scala:151]: Seems somehow only a single column is supported in `BucketTransform`. Will fix this.

**Comment by huaxingao** [comment]
> It's better to update the title and description accordingly before merging.

Updated. Thanks!

**Review by cloud-fan — APPROVED** [review]
LGTM if tests pass

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Comment by huaxingao** [comment]
Thanks!

**Pull request closed without merging**
